plugins/single_plugins/fast-switcher.cpp

```diff
@@ -0,0 +1,194 @@
+#include <output.hpp>
+#include <debug.hpp>
+#include <opengl.hpp>
+#include <signal-definitions.hpp>
+#include <view.hpp>
+#include <workspace-manager.hpp>
+#include <linux/input-event-codes.h>
+
+/*
+ * This plugin rovides abilities to switch between views.
+ * It works similarly to the alt-esc binding in Windows or GNOME
+ */
+
+class fast_switcher : public wayfire_plugin_t
+{
+    key_callback init_binding;
+    wf_option activate_key;
+
+    signal_callback_t destroyed;
+
+    struct
+    {
+        bool active = false;
+
+        bool mod_released = false;
+        bool in_fold = false;
+        bool in_unfold = false;
+        bool in_rotate = false;
+
+        bool reversed_folds = false;
+
+        bool in_continuous_switch = false;
+        bool in_switch = false;
+    } state;
+
+    size_t current_view_index;
+
+    std::vector<wayfire_view> views; // all views on current viewport
+
+    public:
+
+    void init(wayfire_config *config)
+    {
+        grab_interface->name = "fast_switcher";
+        grab_interface->abilities_mask = WF_ABILITY_CONTROL_WM;
+
+        auto section = config->get_section("fast_switcher");
+
+        activate_key = section->get_option("activate", "<alt> KEY_TAB");
+        init_binding = [=] (uint32_t key)
+        {
+            fast_switch();
+        };
+
+        output->add_key(activate_key, &init_binding);
+
+        using namespace std::placeholders;
+        grab_interface->callbacks.keyboard.key = std::bind(std::mem_fn(&fast_switcher::handle_key),
+                this, _1, _2);
+
+        grab_interface->callbacks.keyboard.mod = std::bind(std::mem_fn(&fast_switcher::handle_mod),
+                this, _1, _2);
+    }
+
+    void stop_continuous_switch()
+    {
+        state.in_continuous_switch = false;
+        if (state.in_switch)
+            switch_terminate();
+    }
+
+    void handle_mod(uint32_t mod, uint32_t st)
+    {
+        bool mod_released = (mod == activate_key->as_cached_key().mod && st == WLR_KEY_RELEASED);
+
+        if (mod_released && state.in_switch)
```

Comment:
you can remove the `state.in_switch` altogether because if we are in this function, there's an active grab, so we're switching